miio/viomivacuum.py

```diff
@@ -268,124 +347,340 @@ def has_map(self) -> bool:
+        # - roomIds.length
+        # - *room_id_list
+        # 3rd param of set_mode_withroom is room_array_len and next are
+        # room ids ([0, 1, 3, 11, 12, 13] = start cleaning rooms 11-13).
+        # room ids are encoded in map and it's part of cloud api so best way
+        # to get it is log between device <> mi home app
+        # (before map format is supported).
+        self._cache["edge_state"] = self.get_properties(["mode"])
+        self.send("set_mode_withroom", self._cache["edge_state"] + [1, 0])
 
-    @command()
-    def stop(self):
-        """Stop cleaning."""
-        self.send("set_mode", [0])
+    @command(
+        click.option(
+            "--rooms",
+            "-r",
+            multiple=True,
+            help="Rooms name or room id. Can be used multiple times",
+        )
+    )
+    def start_with_room(self, rooms):
+        """Start cleaning specific rooms."""
+        if not self._cache["rooms"]:
+            self.get_rooms()
+        reverse_rooms = {v: k for k, v in self._cache["rooms"].items()}
+        room_ids = []
+        for room in rooms:
+            if room in self._cache["rooms"]:
+                room_ids.append(room)
+            elif room in reverse_rooms:
+                room_ids.append(reverse_rooms[room])
+            else:
+                return "Rooms {} is unknown, it should be in '{}' " "or in '{}'".format(
+                    room,
+                    ", ".join(self._cache["rooms"].keys()),
+                    ", ".join(self._cache["rooms"].values()),
+                )
+        self._cache["edge_state"] = self.get_properties(["mode"])
+        self.send(
+            "set_mode_withroom",
+            self._cache["edge_state"] + [1, 0, len(room_ids)] + room_ids,
+        )
 
     @command()
     def pause(self):
         """Pause cleaning."""
-        self.send("set_mode_withroom", [0, 2, 0])
+        # params: [edge_state, 0]
+        # - edge: see ViomiEdgeState
+        # - 2: pause cleaning
+        if not self._cache["edge_state"]:
+            self._cache["edge_state"] = self.get_properties(["mode"])
+        self.send("set_mode", self._cache["edge_state"] + [2])
+
+    @command()
+    def stop(self):
+        """Validate that Stop cleaning."""
+        # params: [edge_state, 0]
+        # - edge: see ViomiEdgeState
+        # - 0: stop cleaning
+        if not self._cache["edge_state"]:
+            self._cache["edge_state"] = self.get_properties(["mode"])
+        self.send("set_mode", self._cache["edge_state"] + [0])
+
+    @command(click.argument("mode", type=EnumType(ViomiMode)))
+    def clean_mode(self, mode: ViomiMode):
+        """Set the cleaning mode.
+
+        [vacuum, vacuumAndMop, mop, cleanzone, cleanspot]
+        """
+        self.send("set_mop", [mode.value])
 
     @command(click.argument("speed", type=EnumType(ViomiVacuumSpeed)))
     def set_fan_speed(self, speed: ViomiVacuumSpeed):
-        """Set fanspeed [silent, standard, medium, turbo]."""
```

Comment:
Why change this? I think it is not too long to be shown in the cli tool, which can be helpful?

Reply:
fixed